### PR TITLE
fix: add v2 for getServices and GetTopOperations

### PR DESCRIPTION
**Pull request by nityanandagohain**
Fixes https://github.com/SigNoz/signoz/issues/6514

* Using the resource query builder package for building the resource query.
* Add the v2 functions and replaced the original functions with previous code

----

> [!IMPORTANT]
> Add v2 functions for `GetServices` and `GetTopOperations` in `ClickHouseReader` using a new resource query builder, and update original functions to use v2 when the new schema is enabled.
> 
>   - **Behavior**:
>     - Add `GetServicesV2` and `GetTopOperationsV2` to `ClickHouseReader` for improved query handling using the resource query builder.
>     - Update `GetServices` and `GetTopOperations` to use v2 functions when `useTraceNewSchema` is enabled.
>   - **Functions**:
>     - Add `buildResourceSubQuery()` in `ClickHouseReader` to construct resource sub-queries.
>   - **Constants**:
>     - Remove `TraceResourceBucketFilterWithServiceName` from `constants.go`.
>   - **Utils**:
>     - Add support for `[]string` in `ValidateAndCastValue()` in `format.go`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=SigNoz%2Fsignoz&utm_source=github&utm_medium=referral)<sup> for 296fd2644117dde9866cbc2509b4351f25cb62d3. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: pkg/query-service/app/clickhouseReader/reader.go [hunk 1]
```diff
@@ -552,7 +553,49 @@ func (r *ClickHouseReader) GetTopLevelOperations(ctx context.Context, skipConfig
...
+		}
+		filterSet.Items = append(filterSet.Items, it)
+	}
+	filterSet.Items = append(filterSet.Items, v3.FilterItem{
+		Key: v3.AttributeKey{
+			Key:      "service.name",
+			DataType: v3.AttributeKeyDataTypeString,
+			Type:     v3.AttributeKeyTypeResource,
+		},
+		Operator: v3.FilterOperatorContains,
```
Comment (srikanthccv) [inline pkg/query-service/app/clickhouseReader/reader.go:578]: Why is this contains instead of =?

Comment (nityanandagohain) [reply pkg/query-service/app/clickhouseReader/reader.go]: updated

File: pkg/query-service/app/clickhouseReader/reader.go [hunk 2]
```diff
@@ -552,7 +553,48 @@ func (r *ClickHouseReader) GetTopLevelOperations(ctx context.Context, skipConfig
 	return &operations, nil
 }
 
-func (r *ClickHouseReader) GetServices(ctx context.Context, queryParams *model.GetServicesParams, skipConfig *model.SkipConfig) (*[]model.ServiceItem, *model.ApiError) {
+func (r *ClickHouseReader) buildResourceSubQuery(tags []model.TagQueryParam, svc string, start, end time.Time) (string, error) {
+	filterSet := v3.FilterSet{}
+	for _, tag := range tags {
```
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply pkg/query-service/app/clickhouseReader/reader.go:560]: makes sense

File: pkg/query-service/app/clickhouseReader/reader.go [hunk 3]
```diff
@@ -552,7 +553,48 @@ func (r *ClickHouseReader) GetTopLevelOperations(ctx context.Context, skipConfig
...
+		dataType := v3.AttributeKeyDataTypeString
+		it := v3.FilterItem{
+			Key: v3.AttributeKey{
+				Key:      tag.Key,
+				DataType: dataType,
+				Type:     v3.AttributeKeyTypeResource,
+			},
+			Operator: v3.FilterOperator(tag.Operator),
+			Value:    tag.StringValues[0],
```
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply pkg/query-service/app/clickhouseReader/reader.go]: Updated the builder, confirmed with frontend team that only `in` and `notin` is supported. So added support for them

**Review by srikanthccv — APPROVED** [review]

**Pull request merged**